add mkldnn pooling backward

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37218 BFloat16: enable prepacked weights's inference
* #37215 BFloat16: add explicit dtype support for to_mkldnn and to_dense
* #37199 add mkldnn relu backward and reshape backward
* #37147 add mkldnn batch_norm backward
* **#37146 add mkldnn pooling backward**
* #36122 add mkldnn linear backward
* #36121 enable mkldnn conv2d backward to support mkldnn tensor input

Differential Revision: [D22440968](https://our.internmc.facebook.com/intern/diff/D22440968)

XiaobingSuper committed:
add mkldnn pooling backward

[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn pooling backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn pooling backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn pooling backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn pooling backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn pooling backward"




[ghstack-poisoned]

albanD commented on aten/src/ATen/native/mkldnn/Pooling.cpp:
```diff
@@ -226,10 +346,100 @@ Tensor& mkldnn_adaptive_avg_pool2d_out(
+Tensor mkldnn_adaptive_avg_pool2d_backward(
+    const Tensor& grad_output,
+    const Tensor& input) {
+  TORCH_CHECK(input.dim() == 4, "mkldnn_adaptive_avg_pool2d_backward: Expect 2D input");
```
nit: maybe make the error message slightly more explicit that a 4D Tensor is expected.

XiaobingSuper replied:
@albanD, the file use many **AT_ASSERTM**, any difference with **TORCH_CHECK**?  **AT_ASSERTM** will be deprecated, may can replace them.

albanD replied:
Yes you can replace all the AT_ASSERT with TORCH_CHECK

albanD commented on aten/src/ATen/native/mkldnn/Pooling.cpp:
```diff
@@ -226,10 +346,100 @@ Tensor& mkldnn_adaptive_avg_pool2d_out(
+    TORCH_CHECK(s2 != 0, "output size can not be zero");
+    TORCH_CHECK(
+        s1 % s2 == 0,
+        "input size is not divisible by the output size is not supported yet");
```
nit: this sentence might be clearer: "The given input size is not divisible by the output size. This is not supported yet for mkldnn Tensors".

albanD commented on aten/src/ATen/native/mkldnn/Pooling.cpp:
```diff
@@ -144,6 +192,78 @@ static Tensor _mkldnn_pool2d(
+    // adjust padding until output sizes agree
+    bool all_equal = false;
+    std::vector<int64_t> output_sizes;
+    while (!all_equal) {
```
Given that all ops are calling here. How expensive is this? Should there be a warning in the docs that the ceil_mode might be inefficient for mkldnn Tensors?

albanD commented on aten/src/ATen/native/mkldnn/Pooling.cpp:
```diff
@@ -17,8 +17,7 @@ Tensor mkldnn_max_pool2d(
     bool ceil_mode) {
-  AT_ERROR(
-      "mkldnn_max_pool2d: ATen not compiled with MKLDNN support");
+  TORCH_CHECK(false, "mkldnn_max_pool2d: ATen not compiled with MKLDNN support");
```
Thanks for taking the time to fix these.

albanD reviewed (approved):
Looks good, small comments only.

XiaobingSuper committed:
Update on "add mkldnn pooling backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn pooling backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn pooling backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn pooling backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn pooling backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn pooling backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn pooling backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn pooling backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn pooling backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn pooling backward"


Differential Revision: [D22440968](https://our.internmc.facebook.com/intern/diff/D22440968)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn pooling backward"


Differential Revision: [D22440968](https://our.internmc.facebook.com/intern/diff/D22440968)

[ghstack-poisoned]
